drivers/i2c/i2c_wch.c

```diff
@@ -0,0 +1,498 @@
+		wch_i2c_handle_txe(dev);
+	} else {
+		wch_i2c_handle_rxne(dev);
+	}
+}
+
+static void i2c_wch_event_isr(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+	uint16_t status = regs->STAR1;
+	bool write;
+
+	write = ((data->current.msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE);
+
+	if (status & I2C_STAR1_SB) {
+		wch_i2c_handle_start_bit(dev);
+	} else if (status & I2C_STAR1_ADDR) {
+		wch_i2c_handle_addr(dev);
+	} else if (status & I2C_STAR1_BTF) {
+		wch_i2c_handle_btf(dev);
+	} else if ((status & I2C_STAR1_TXE) && write) {
+		wch_i2c_handle_txe(dev);
+	} else if ((status & I2C_STAR1_RXNE) && (!write)) {
+		wch_i2c_handle_rxne(dev);
+	}
+}
+
+static void i2c_wch_error_isr(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+	uint16_t status = regs->STAR1;
+
+	if (status & (I2C_STAR1_AF | I2C_STAR1_ARLO | I2C_STAR1_BERR)) {
+		if (status & I2C_STAR1_AF) {
+			regs->CTLR1 |= I2C_CTLR1_STOP;
+		}
+
+		data->current.err |= ((status >> 8) & 0x7);
+
+		regs->STAR1 &= ~(I2C_STAR1_AF | I2C_STAR1_ARLO | I2C_STAR1_BERR);
+
+		atomic_set(&data->xfer_done, 1);
+	}
+}
+
+static void wch_i2c_msg_init(const struct device *dev, struct i2c_msg *msg, uint16_t addr)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	atomic_set(&data->xfer_done, 0);
+
+	data->current.msg = msg;
+	data->current.idx = 0U;
+	data->current.err = 0U;
+	data->current.addr = addr;
+
+	regs->CTLR1 |= I2C_CTLR1_PE;
+	regs->CTLR1 |= I2C_CTLR1_ACK;
+
+	if (msg->flags & I2C_MSG_RESTART) {
+		if (regs->CTLR1 & I2C_CTLR1_STOP) {
+			regs->CTLR1 &= ~I2C_CTLR1_STOP;
+		}
+		regs->CTLR1 |= I2C_CTLR1_START;
+	}
+}
+
+static int32_t wch_i2c_msg_end(const struct device *dev)
+{
+	struct i2c_wch_data *data = dev->data;
+
+	if (!data->current.err) {
+		return 0;
+	}
+
+	if (data->current.err & (I2C_STAR1_ARLO >> 8)) {
+		LOG_DBG("ARLO");
+	}
+
+	if (data->current.err & (I2C_STAR1_AF >> 8)) {
+		LOG_DBG("NACK");
+	}
+
+	if (data->current.err & (I2C_STAR1_BERR >> 8)) {
+		LOG_DBG("ERR");
+	}
+
+	data->current.err = 0;
+
+	return -EIO;
+}
+
+static void wch_i2c_config_interrupts(I2C_TypeDef *regs, bool enable)
+{
+	if (enable) {
+		regs->CTLR2 |= (I2C_CTLR2_ITERREN | I2C_CTLR2_ITEVTEN | I2C_CTLR2_ITBUFEN);
+	} else {
+		regs->CTLR2 &= ~(I2C_CTLR2_ITERREN | I2C_CTLR2_ITEVTEN | I2C_CTLR2_ITBUFEN);
+	}
+}
+
+static int32_t wch_i2c_begin_transfer(const struct device *dev, struct i2c_msg *msg, uint16_t addr)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	wch_i2c_msg_init(dev, msg, addr);
+
+	wch_i2c_config_interrupts(regs, true);
+
+	k_sched_lock();
+	while (atomic_get(&data->xfer_done) == 0) {
```

Comment:
Can this yield in some way? It seems a pity to block the CPU while waiting ~1 ms for an I2C transfer to complete.

Reply:
Thank you so much for the review, I will do all the changes on sunday/monday. Regarding this, it was still a WIP and I will change it.

Reply:
Outdated: I'm using a semaphone now.